src/topology/algebra/uniform_extension.lean

```diff
@@ -0,0 +1,241 @@
+/-
+Copyright (c) 2019 Zhouhang Zhou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Zhouhang Zhou
+-/
+
+import topology.uniform_space.uniform_embedding
+import topology.algebra.module
+import analysis.normed_space.operator_norm
+
+/-!
+# Extension of uniform continuous functions
+
+Extension of a uniform continuous function preserves its properties.
+
+## Main statements
+
+Let `β` be a dense subset of `α`, with `e : β → α` as the inclusion map. Let `f : β → γ` be a
+uniform continuous function on `β`. Then there is a uniform continuous function `ψ : α → γ` on `α`,
+such that `ψ` is an extension of `f`, i.e. `ψ ∘ e = f`.
+
+This theorem is proved in `topology.uniform_space.uniform_embedding.lean`. See the lemma
+`uniform_continuous_uniformly_extend`.
+
+In this file we show that
+
+* If `f` is an `add_monoid_hom`, then `ψ` is an `add_monoid_hom`
+* If `f` is a `continuous_linear_map`, them `ψ` is a `continuous_linear_map`
+* If `∥f∥ ≤ M`, then `∥ψ∥ ≤ M`
+
+## Implementation notes
+
+`e` is generalized to be `uniform_inducing`.
+
+## Tags
+
+uniform continuous, extend, extension, operator norm, op_norm
+
+-/
+
+noncomputable theory
+
+open set filter lattice function
+
+open_locale classical
+
+variables {α : Type*} {β : Type*} {γ : Type*}
+
+section uniformly_extend_zero
+
+variables [uniform_space α] [has_zero α]
+variables [uniform_space β] [has_zero β]
+variables [uniform_space γ] [has_zero γ] [separated γ]
+variables (e : β → α) (e_zero : e 0 = 0) (h_e : uniform_inducing e) (h_dense : dense_range e)
+variables {f : β → γ} (h_f : uniform_continuous f)
+
+include e_zero h_f
+
+local notation `ψ` := (h_e.dense_inducing h_dense).extend f
+
+lemma uniformly_extend_zero (h : f 0 = 0) : ψ 0 = 0 :=
+by { rw ← e_zero, rwa uniformly_extend_of_ind h_e h_dense h_f }
+
+end uniformly_extend_zero
+
+section uniformly_extend_add
+
+variables [uniform_space α] [add_monoid α] [topological_add_monoid α]
+variables [uniform_space β] [add_monoid β]
+variables [uniform_space γ] [add_monoid γ] [topological_add_monoid γ] [separated γ]
+          [complete_space γ]
+variables (e : β → α) (e_add : ∀b₁ b₂, e (b₁ + b₂) = e b₁ + e b₂) (h_e : uniform_inducing e)
+          (h_dense : dense_range e)
+variables {f : β → γ} (h_f : uniform_continuous f)
+
+include e_add h_f
+
+local notation `ψ` := (h_e.dense_inducing h_dense).extend f
+
+lemma uniformly_extend_add (h : ∀ b₁ b₂, f (b₁ + b₂) = f b₁ + f b₂) :
+  ∀ a₁ a₂, ψ (a₁ + a₂) = ψ a₁ + ψ a₂ :=
```

Comment:
Shouldn't you prove this for general `ψ` that agree with `f` when restricted to `β`? In other words, I think you should use the characteristic predicate of an extended function, instead of the explicit construction. This might also apply elsewhere.
Besides that, note that a lot of effort was spent on ring completions in the perfectoid project. Have you looked at https://github.com/leanprover-community/lean-perfectoid-spaces/tree/master/src/for_mathlib/uniform_space ?

Reply:
Group and ring completions are already in mathlib. 
```
lemma is_add_group_hom_extension {f : α → β} [is_add_group_hom f] (hf : continuous f) : is_add_group_hom (completion.extension f)
```
I think this one is what I need, in fact with the same proof I think, but the extended function is from `completion α` to `β`, so I'm not sure how to use it for my purpose where an inclusion map is given...